internal/backend/post.go

```diff
@@ -46,64 +48,74 @@ func (d *delphisBackend) CreatePost(ctx context.Context, discussionID string, pa
 		ImportedContentID: input.ImportedContentID,
 	}
 
-	// Begin tx
-	tx, err := d.db.BeginTx(ctx)
-	if err != nil {
-		logrus.WithError(err).Error("failed to begin tx")
-		return nil, err
-	}
-	// Put post contents
-	if err := d.db.PutPostContent(ctx, tx, postContent); err != nil {
-		logrus.WithError(err).Error("failed to PutPostContent")
-
-		// Rollback on errors
-		if txErr := d.db.RollbackTx(ctx, tx); txErr != nil {
-			logrus.WithError(txErr).Error("failed to rollback tx")
-			return nil, multierr.Append(err, txErr)
+	retryAttempts := 0
+	for true {
+		// Begin tx
+		tx, err := d.db.BeginTx(ctx)
+		if err != nil {
+			logrus.WithError(err).Error("failed to begin tx")
+			return nil, err
 		}
-		return nil, err
-	}
-
-	// Put post
-	postObj, err := d.db.PutPost(ctx, tx, post)
-	if err != nil {
-		logrus.WithError(err).Error("failed to PutPost")
-
-		// Rollback on errors
-		if txErr := d.db.RollbackTx(ctx, tx); txErr != nil {
-			logrus.WithError(txErr).Error("failed to rollback tx")
-			return nil, multierr.Append(err, txErr)
+		// Put post contents
+		if err := d.db.PutPostContent(ctx, tx, postContent); err != nil {
+			logrus.WithError(err).Error("failed to PutPostContent")
+
+			// Rollback on errors
+			if txErr := d.db.RollbackTx(ctx, tx); txErr != nil {
+				logrus.WithError(txErr).Error("failed to rollback tx")
+				return nil, multierr.Append(err, txErr)
+			}
+			return nil, err
 		}
-		return nil, err
-	}
 
-	// Put Activity
-	if err := d.db.PutActivity(ctx, tx, postObj); err != nil {
-		logrus.WithError(err).Error("failed to PutActivity")
+		// Put post
+		postObj, err := d.db.PutPost(ctx, tx, post)
+		if err != nil {
+			if retryAttempts < PutPostMaxRetry && strings.Contains(strings.ToLower(err.Error()), "duplicate key value violates unique constraint") {
+				retryAttempts++
+				logrus.WithError(err).Error("failed to PutPost, retrying with attempt #" + strconv.Itoa(retryAttempts))
+				// Note for the future: hould we backoff a little?
```

Comment:
Not a huge concern for now, but if a putpost fail it may be due to excessive traffic on this operation, namely too many clients trying to create a post in the same instance. Backing off will decrease the probability of continuous failures, but I think this is not a priority now.

Reply:
Seems reasonable. The idea of exponential backoff is reasonable to consider.